README.md
# Life Calendar

This plugin renders yaml-based fenced code into a Life Calendar inspired by Tim Urban's article "Your Life in Weeks".

![screenshot](https://raw.githubusercontent.com/hieuthi/joplin-plugin-life-calendar/main/docs/life-calendar-v1.2.0-screenshot.png)

The purpose of Life Calendar is documenting the events in your life but there is nothing stopping you from using it for other purposes such as project management. The yaml-based syntax make it easy to export data to other format.

## Usage
You need to create a fenced code with `life` as language to render the calendar then input the events using the the follow template:
```
  - date : yyyy-mm-dd    # (required) (interpretation DD/MM/YYYY 12:00 local time)
    title:               # (required) event title
    icon :               # (optional) event icon, use 1st character of title if icon is null
    z-index:             # (optional) priority of the event, the one with the bigest value will be shown (default:0)
    className:           # (optional) css class of the event (color, background-color, etc.)
    color:               # (optional) overwrite color
    backgroundColor:     # (optional) overwrite background-color

```
This plugin also support periods which is a span of time instead of a point in time. The template for period is as follow:
```
  - start: yyyy-mm-dd    # (required) (interpretation: DD/MM/YYYY 01:00 local time)
    end  : yyyy-mm-dd    # (required) (interpretation: DD/MM/YYYY 23:00 local time)
    title:               # (required) event title
    className:           # (optional) css class of the event (color, background-color, etc.)
    color:               # (optional) overwrite color
    icon:                # (optional) period icon
    backgroundColor:     # (optional) overwrite background-color
```

The example below should demostrate all the supported features. It is recommended to intend the events block with spaces as yalm parser is quite sensitive. You may encounter parsing problems when typing with Joplin as it use tab for intending.

From v1.3.0 forward, this plugin also support day-based, month-based, and year-based calendar beside week-based by writting down the desired type `type: month`. Moreover you can click on the a square to pin down the details window so you can read. Click again at any square to unpin it.

`````markdown
# 📆 Life Calendar

<style>
  .education { color:white !important; background-color:red !important; } 
</style>

- **Date of Birth**: 01/01/2020

```life
type: week        # [ day, week, month, year ]
dob : 2020-01-01  # alias=startDate
lifespan: 4       # alias=duration
# dod: 2026-01-01 # alias=endDate
ages: [1,2,3,4,5,6]
events:
  - { date: 2022-04-19, title: Inline Event, icon: 🍿, backgroundColor: navy, color: white }
  - date : 2021-06-12
    title: Two events in the same week
    color: yellow
  - date : 2021-06-10
    title: Got a Gold Medal
    icon : 🥇
    z-index: 1 # shown icon of this event in calendar board instead of the other one
    backgroundColor: green
  - date : 2021-03-28
    title: Random Event
    icon : 🎤
    color: black
    backgroundColor: '#6495ED'
  - date : 2021-01-03
    title: Birthday Party
    icon : 🎂
    color: black
    backgroundColor: hotpink
  - date : 2020-08-19
    title: Obtain Magician License
    icon : 🎩
    className: education
  - date : 2020-03-19
    title: First character used as icon
    backgroundColor: orange
    color: black
  - date : 2020-03-27
    title: Used explicit icon
    icon : '<b>F</b>'
    color: orange
    backgroundColor: black
  - date : 2020-01-01
    title: Date of Birth
    icon : 👶
    backgroundColor: transparent
periods:
  - start: 2022-09-19
    end  : 2022-11-05
    title: Sunny
    icon : 🌤
  - start: 2022-02-16
    end  : 2022-08-16
    title: Advanced Magic Course
    className: education
  - start: 2020-12-15
    end: 2022-01-01
    title: High school
    color: black
    backgroundColor: lightblue
  - start: 2020-02-01
    end: 2020-12-15
    title: Preschool
    backgroundColor: yellow
    color: black
```

`````

You can further customize the apperance using `userstyle.css`

## Tips & Tricks
### Ages template
Some pre-prepared templates for `ages` option, just copy it to your note instead of typing yourself:
```
ages: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]
ages: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98]
ages: [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99]
ages: [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96]
ages: [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95]
```

## Acknowledgements
- Tim Urban for popularizing the [Your Life in Weeks Concept](https://waitbutwhy.com/2014/05/life-weeks.html)
- [Dylan N's Life Calendar Implementation](https://github.com/ngduc/life-calendar) which is a reference for this plugin.

## License
[MIT](https://raw.githubusercontent.com/hieuthi/joplin-plugin-life-calendar/main/LICENSE)

CHANGELOG.md
# Changelog

## [v1.4.1] - 2023-02-05
### Added
- Add icon support for period @shubhamjain0594

## [v1.4.0] - 2022-12-23
### Changed
- Change period end date intepretation to inclusive 
- Strengthen datetime handling
- Change CSS

### Added
- Add className support for period
- Add z-index (priority) support for event


## [v1.3.0] - 2022-12-16
### Changed
- Adjust CSS

### Added
- Support day-base, month-based, and year-based
- Support pin down event window
- Add CodeMirror Mode for life

## [v1.2.1] - 2022-02-14
### Changed
- Change theme

### Added
- Support periods

## [v1.1.0] - 2022-02-06
### Changed
- Render asynchronously

### Added
- Support className

## [v1.0.0] - 2022-02-05
Initial release, fully functional

[v1.4.1]: https://github.com/hieuthi/joplin-plugin-life-calendar/compare/v1.4.0...v1.4.1
[v1.4.0]: https://github.com/hieuthi/joplin-plugin-life-calendar/compare/v1.3.0...v1.4.0
[v1.3.0]: https://github.com/hieuthi/joplin-plugin-life-calendar/compare/v1.2.1...v1.3.0
[v1.2.1]: https://github.com/hieuthi/joplin-plugin-life-calendar/compare/v1.1.0...v1.2.1
[v1.1.0]: https://github.com/hieuthi/joplin-plugin-life-calendar/compare/v1.0.0...v1.1.0
[v1.0.0]: https://github.com/hieuthi/joplin-plugin-life-calendar/releases/tag/v1.0.0

src/lifeMdRule.js
const YAML = require('yaml');

module.exports = {
  default: function(context) {
    return {
      plugin: function (markdownIt, _options) {
        const defaultRender = markdownIt.renderer.rules.fence || function(tokens, idx, options, env, self) {
          return self.renderToken(tokens, idx, options, env, self);
        };
        markdownIt.renderer.rules.fence = function(tokens, idx, options, env, self) {
          const token = tokens[idx];
          if (token.info !== 'life') return defaultRender(tokens, idx, options, env, self);
          try {
            var contentHtml = JSON.stringify(YAML.parse(markdownIt.utils.escapeHtml(token.content)));
          } catch (e) {
            var contentHtml = {};
          }
          return `
            <div class="joplin-editable">
              <div class="life-calendar">${contentHtml}</div>
            </div>
          `;
        };
      },
      assets: function () {
        return [
          { name: 'life-calendar.js' },
          { name: 'life-calendar.css'},]
      }
    }
  }
}

src/life-calendar.js
document.addEventListener('joplin-noteDidUpdate', buildCalendars );

var _life_pin = null;

if (/WebKit/i.test(navigator.userAgent)) { // sniff
    var _timer_life = setInterval(function() {
        if (/loaded|complete/.test(document.readyState)) {
            buildCalendars()
        }
    }, 10);
}

// Helpers functions
function dateToIsoString(date){
  var month = (date.getMonth()+1).toString();
  var day   = (date.getDate()).toString()
  month = month.length == 1 ? '0' + month : month;
  day   = day.length == 1 ? '0' + day : day;
  return `${date.getFullYear()}-${month}-${day}`
}

function isoStringToDate(iso) {
  if (iso==null) { return null }
  var [yyyy, mm, dd] = iso.split('-');
  return new Date(yyyy, mm - 1, dd); 
}

function buildCalendars() {
  if (_timer_life) clearInterval(_timer_life);
  _life_pin = null;

  const calendars = document.getElementsByClassName('life-calendar');
  for (var i=0; i<calendars.length; i++){
    var calendar = calendars[i];
    try {
      var options = JSON.parse(calendar.textContent);
      setTimeout(makeLifeCalendar(calendar, options),60);
    } catch (e) {
      calendar.innerHTML = "Parsing Error";
      console.log("Error: " + e);
    }
  }
}

function renderInfo(infoElement, info) {
  var elem = document.createElement("div");
  elem.innerHTML = info["header"];
  elem.className = 'info-meta';
  infoElement.appendChild(elem);

  info["events"].forEach(event => {
    var elem = document.createElement("div");
    elem.className = 'info-event';
    elem.innerHTML = `<b>${event["date"]}</b> <span class="icon">${event["icon"] || event["title"][0]}</span> ${event["title"]}`
    if (event["className"]){ elem.className = elem.className + " " + event["className"]; }
    if (event["color"]){ elem.style.color = event["color"]; }
    if (event["backgroundColor"]){ elem.style.backgroundColor = event["backgroundColor"]; }
    infoElement.appendChild(elem);
  });
  info["periods"].forEach(period => {
    var elem = document.createElement("div");
    elem.className = 'info-period';
    elem.innerHTML = `<b>${period["sIso"]}</b> <b>${period["eIso"]}</b>`;
    if (period["icon"]){elem.innerHTML = elem.innerHTML + ` <span class="icon">${period["icon"]}</span>`; }
    elem.innerHTML = elem.innerHTML + ` ${period["title"]}`;
    if (period["className"]){ elem.className = elem.className + " " + period["className"]; }
    if (period["color"]){ elem.style.color = period["color"]; }
    if (period["backgroundColor"]){ elem.style.backgroundColor = period["backgroundColor"]; }
    infoElement.appendChild(elem);
  })
}

function itemMouseOut() {
  var firstInfo  = this.parentElement.parentElement.firstChild;
  var secondInfo = firstInfo.nextSibling;
  secondInfo.classList.remove('visible')
  if (_life_pin !== null){
    firstInfo.classList.add('visible')
  }
}
function itemMouseOver() {
  var firstInfo  = this.parentElement.parentElement.firstChild;
  var secondInfo = firstInfo.nextSibling;
  secondInfo.innerHTML = '';
  secondInfo.classList.add('visible')
  renderInfo(secondInfo, this.infoObj);
  if (_life_pin !== null){
    firstInfo.classList.remove('visible');
  }
}
function itemClick() {
  var firstInfo  = this.parentElement.parentElement.firstChild;
  if (_life_pin == null){
    _life_pin = this.infoObj;
    firstInfo.innerHTML = '';
    firstInfo.classList.add('visible')
    renderInfo(firstInfo, this.infoObj);
  } else {
    firstInfo.classList.remove('visible');
    _life_pin = null;
  }
}

const MSHOUR = 3.6e+6;
const MSDAY  = MSHOUR*24;
const MSWEEK = MSDAY * 7;
function getEventIndex(ctype, startDate, eventDate) {
  var idx = 0;
  switch (ctype) {
    case "day":
      idx = Math.floor((eventDate.getTime()-startDate.getTime()) / MSDAY);
      break;
    case "week":
      idx = Math.floor((eventDate.getTime()-startDate.getTime()) / MSWEEK);
      break;
    case "month":
      idx = (eventDate.getFullYear() - startDate.getFullYear()) * 12 + (eventDate.getMonth() - startDate.getMonth());
      break;
    case "year":
      idx = eventDate.getFullYear() - startDate.getFullYear();
      break;
  }
  return idx;
}
function getEventTime(ctype, startDate, index) {
  var date = startDate;
  switch (ctype) {
    case "day":
      date = new Date(startDate.getTime() + index*MSDAY);
      break;
    case "week":
      date = new Date(startDate.getTime() + index*MSWEEK);
      break;
    case "month":
      var month = startDate.getMonth() + index;
      date = new Date(startDate.getFullYear() + Math.floor(month/12), month % 12, startDate.getDate());
      break;
    case "year":
      date = new Date(startDate.getFullYear() + index, startDate.getMonth(), startDate.getDate());
      break;
  }
  return date;
}

function floorEventDate(ctype, date) {
  var ret = new Date(date.getFullYear(), date.getMonth(), date.getDate());
  switch (ctype) {
    case "day":
      break;
    case "week":
      ret = new Date(ret.getTime() - ret.getDay()*MSDAY)
      break;
    case "month":
      ret = new Date(ret.getFullYear(), ret.getMonth(), 1);
      break;
    case "year":
      ret = new Date(ret.getFullYear(), 0, 1);
      break;
  }
  return ret;
}

function makeLifeCalendar(calendar, options) {
  calendar.innerHTML = ''; // Clear element content

  var firstInfo = document.createElement("div");
  firstInfo.className = 'info';
  calendar.appendChild(firstInfo);
  calendar.appendChild(firstInfo.cloneNode());

  var ctype = options["type"] || "week";
  ctype = !["day", "week", "month", "year"].includes(ctype) ? "week" : ctype;
  var itype = "W"
  switch (ctype) {
    case "day"  : itype = "D"; break;
    case "week" : itype = "W"; break;
    case "month": itype = "M"; break;
    case "year" : itype = "Y"; break;
  }

  // Calculate time period

  var dob  = options["dob"]  ? isoStringToDate(options["dob"])       : isoStringToDate("1970-01-01");
  dob = options["startDate"] ? isoStringToDate(options["startDate"]) : dob;
  var [ byear, bmonth, bdate ] = [ dob.getFullYear(), dob.getMonth(), dob.getDate() ];
  var dodm = getEventTime(ctype, dob, 9999);

  var dod = options["lifespan"] ? new Date(byear+options["lifespan"],bmonth,bdate) : new Date(byear+5,bmonth,bdate);
  dod = options["duration"]     ? new Date(byear+options["duration"],bmonth,bdate) : dod;
  dod = options["dod"]          ? isoStringToDate(options["dod"])                  : dod;
  dod = options["endDate"]      ? isoStringToDate(options["endDate"])              : dod;
  dod = dod > dodm ? dodm : dod;

  const doba = floorEventDate(ctype, dob);
  const doda = floorEventDate(ctype, dod);
  const now  = new Date();

  // Prepare ages
  options["events"] = options["events"] || [];
  options["events"].push({
    "date" : dateToIsoString(now),
    "title": `Today`,
    "icon" : `>`
  });
  if (options["ages"]) {
    options["ages"].forEach(age => {
      var birthday = null;
      if (bmonth==1 && bdate==29 && age % 4 != 0){
        birthday = new Date(byear+age,1,28);
      } else {
        birthday = new Date(byear+age,bmonth,bdate);
      }
      options["events"].push({
        "date" : dateToIsoString(birthday),
        "title": `Turning ${age}`,
        "icon" : `${age}` 
      })
    })
  }

  const events = {}
  // Prepare events
  if (options["events"]) {
    options["events"].sort( (a,b) => {return a["date"] > b["date"] ? -1 : 1; })
    options["events"].forEach(item => {
      var edate = new Date(isoStringToDate(item["date"]).getTime()+MSHOUR*12); // Daylight Saving Time
      var idx   = getEventIndex(ctype, doba, edate);
      if (events[idx]){
        events[idx].push(item);
      } else {
        events[idx] = [item];
      }
    })
  }
  // Prepare periods
  const periods = []
  if (options["periods"]){
    options["periods"].sort( (a,b) => {return a["start"] > b["start"] ? -1 : 1})
    options["periods"].forEach(item => {
      var sDate = new Date(isoStringToDate(item["start"]).getTime() + MSHOUR + 1); // Daylight Saving Time
      var eDate = new Date(isoStringToDate(item["end"]).getTime() + MSHOUR*23 - 1); // Daylight Saving Time
      if (sDate && eDate){
        var title = item["title"];
        var className = item["className"] || null;
        var color = item["color"] || null;
        var backgroundColor = item["backgroundColor"] || null;
        var icon = item["icon"] || null;
        periods.push({"start": sDate, "end": eDate,
          "sIso": dateToIsoString(sDate), "eIso": dateToIsoString(eDate),
          "title": title, "color": color, "backgroundColor": backgroundColor, 
          "icon": icon, "className": className})
      }
    })
  }

  const children = [];

  const nowIdx  = getEventIndex(ctype, doba, now);
  const nItems  = getEventIndex(ctype, doba, doda) + 1;
  for (var i=0; i<nItems; i++){
    var dateStart = new Date(getEventTime(ctype, doba, i).getTime() + MSHOUR + 1);     // Daylight Saving Time
    var dateEnd   = new Date(getEventTime(ctype, doba, i+1).getTime() - MSHOUR - 1); // Daylight Saving Time

    var age = Math.abs((new Date(dateEnd.getTime() - dob.getTime())).getUTCFullYear()-1970); //Daulight Saving Time


    var info = {"events" : events[i] || [],
                "periods": [],
                "header" : `<b>${itype}${i.toString().padStart(4,'0')}</b> | ${dateToIsoString(dateStart)}~${dateToIsoString(dateEnd)} | Age: ${age}`};

    var item     = document.createElement("div");
    var itemspan = document.createElement("span");
    // Past, Present, and Future
    if ( i < nowIdx ) {
      item.className = 'life-item past';
    } else if ( i == nowIdx) {
      item.className = 'life-item present';
    } else {
      item.className = 'life-item future';
    }
    // itemIcon is set by event if available, else by period
    var itemIcon = null;
    // Prepare info events
    if (events[i]){
      var event = events[i][0]; 
      events[i].forEach(item => {
        if ((item["z-index"] || 0) > (event["z-index"] ||0)) {
          event = item;
        }
      })
      itemIcon = event["icon"] || event["title"][0];
      if (event["className"]){ itemspan.className = event["className"]; }
      if (event["color"]){ itemspan.style.color = event["color"]; }
      if (event["backgroundColor"]){ itemspan.style.backgroundColor = event["backgroundColor"]; }
    }

    // Periods
    var bgcolor = null;
    var stclass = null;
    periods.forEach(period => {
      if ( !(dateEnd<=period["start"] || dateStart>=period["end"]) ) {
        info["periods"].push(period);
        if (period["icon"] && itemIcon===null){ itemIcon = period["icon"]; }
        if (period["backgroundColor"] && bgcolor===null){ bgcolor = period["backgroundColor"]; }
        if (period["className"] && stclass===null){ stclass = period["className"]; }
      }
    });
    if (bgcolor){ item.style.backgroundColor = bgcolor; }
    if (stclass){ item.classList.add(stclass) }

    itemspan.innerHTML = itemIcon || "";
    itemspan.infoObj = info;
    itemspan.onmouseover = itemMouseOver;
    itemspan.onmouseout  = itemMouseOut;
    itemspan.onclick     = itemClick;

    item.appendChild(itemspan);
    calendar.appendChild(item);
  }
}